File: packages/domain/src/editorial/article.test.ts
import { describe, expect, it } from 'vitest'
import { NotPermitted } from '../identity/actor.js'
import { articleId, categoryId, familyId, tagId } from '../shared/ids.js'
import { Slug } from '../shared/slug.js'
import { Article } from './article.js'
import { AUTHOR, OTHER, REVISION, actorWith, anApprovedArticle, anArticle } from '../testing/builders.js'
import { IllegalTransition, MissingApprovedRevision, NotOwnArticle, ScheduleInPast } from './errors.js'

const NOW = new Date('2026-08-08T10:00:00Z')
const LATER = new Date('2026-08-08T18:00:00Z')
const EARLIER = new Date('2026-08-08T09:00:00Z')

const author = actorWith(['author'])
const journalist = actorWith(['journalist'])
const editor = actorWith(['editor'], OTHER)
const subscriber = actorWith(['subscriber'], OTHER)

describe('create', () => {
  const input = {
    id: articleId('art_new'),
    familyId: familyId('fam_new'),
    locale: 'en',
    slug: Slug.of('new-piece'),
    title: 'New Piece',
    categoryId: categoryId('cat_news'),
  }

  it('starts life as a draft', () => {
    expect(Article.create(author, input).status).toBe('draft')
  })

  it('attributes authorship to whoever created it, not to a passed-in id', () => {
    expect(Article.create(author, input).authorId).toBe(AUTHOR)
  })

  it('starts with no approval, no schedule and no publication', () => {
    const created = Article.create(author, input).snapshot()
    expect(created.approvedRevisionId).toBeNull()
    expect(created.scheduledAt).toBeNull()
    expect(created.publishedAt).toBeNull()
  })

  it('defaults to no tags', () => {
    expect(Article.create(author, input).snapshot().tagIds).toEqual([])
  })

  it('keeps tags when supplied', () => {
    const tagged = Article.create(author, { ...input, tagIds: [tagId('tag_budget')] })
    expect(tagged.snapshot().tagIds).toEqual(['tag_budget'])
  })

  it('refuses an actor who may not draft', () => {
    expect(() => Article.create(subscriber, input)).toThrow(NotPermitted)
  })
})

describe('submitForReview', () => {
  it('moves a draft into review', () => {
    expect(anArticle().submitForReview(author).status).toBe('in_review')
  })

  it('refuses an actor without the submit permission', () => {
    expect(() => anArticle().submitForReview(subscriber)).toThrow(NotPermitted)
  })

  it("refuses another author's draft", () => {
    const foreign = actorWith(['author'], OTHER)
    expect(() => anArticle().submitForReview(foreign)).toThrow(NotOwnArticle)
  })

  it('reports missing permission before ownership, so it leaks no authorship', () => {
    // A subscriber is both unpermitted and not the author. They must be told
    // the former — "belongs to another author" would reveal who wrote an
    // unpublished draft to someone with no editorial access at all.
    expect(() => anArticle().submitForReview(subscriber)).toThrow(NotPermitted)
  })

  it('allows an editor to submit any draft, because they may edit any', () => {
    expect(anArticle({ authorId: AUTHOR }).submitForReview(editor).status).toBe('in_review')
  })

  it('refuses to submit an article that is already published', () => {
    const published = anArticle({ status: 'published' })
    expect(() => published.submitForReview(author)).toThrow(IllegalTransition)
  })
})

describe('approve and reject', () => {
  it('approve records the revision that was approved', () => {
    const approved = anArticle({ status: 'in_review' }).approve(editor, REVISION)
    expect(approved.status).toBe('approved')
    expect(approved.snapshot().approvedRevisionId).toBe(REVISION)
  })

  it('reject returns the article to draft and clears the approval', () => {
    const rejected = anArticle({ status: 'in_review', approvedRevisionId: REVISION }).reject(editor)
    expect(rejected.status).toBe('draft')
    expect(rejected.snapshot().approvedRevisionId).toBeNull()
  })

  it('a journalist may not approve their own work', () => {
    expect(() => anArticle({ status: 'in_review' }).approve(journalist, REVISION)).toThrow(NotPermitted)
  })
})

describe('publish', () => {
  it('publishes an approved article and stamps the time', () => {
    const published = anApprovedArticle().publish(editor, NOW)
    expect(published.status).toBe('published')
    expect(published.publishedAt).toEqual(NOW)
  })

  it('refuses to publish without an approved revision', () => {
    const noRevision = anArticle({ status: 'approved' })
    expect(() => noRevision.publish(editor, NOW)).toThrow(MissingApprovedRevision)
  })

  it('refuses to publish straight from draft', () => {
    expect(() => anArticle({ approvedRevisionId: REVISION }).publish(editor, NOW)).toThrow(IllegalTransition)
  })

  it('refuses an author, whatever the state', () => {
    expect(() => anApprovedArticle().publish(author, NOW)).toThrow(NotPermitted)
  })

  it('clears the schedule once published', () => {
    const scheduled = anApprovedArticle({ status: 'scheduled', scheduledAt: LATER })
    expect(scheduled.publish(editor, NOW).scheduledAt).toBeNull()
  })

  it('can republish something previously unpublished', () => {
    const pulled = anApprovedArticle({ status: 'unpublished' })
    expect(pulled.publish(editor, NOW).status).toBe('published')
  })
})

describe('schedule', () => {
  it('schedules an approved article for a future moment', () => {
    const scheduled = anApprovedArticle().schedule(editor, LATER, NOW)
    expect(scheduled.status).toBe('scheduled')
    expect(scheduled.scheduledAt).toEqual(LATER)
  })

  it('refuses a moment in the past', () => {
    expect(() => anApprovedArticle().schedule(editor, EARLIER, NOW)).toThrow(ScheduleInPast)
  })

  it('refuses the present moment, which would race the cron', () => {
    expect(() => anApprovedArticle().schedule(editor, NOW, NOW)).toThrow(ScheduleInPast)
  })

  it('refuses to schedule without an approved revision', () => {
    expect(() => anArticle({ status: 'approved' }).schedule(editor, LATER, NOW)).toThrow(MissingApprovedRevision)
  })
})

describe('unpublish', () => {
  it('pulls a published article', () => {
    const published = anApprovedArticle({ status: 'published', publishedAt: NOW })
    expect(published.unpublish(editor).status).toBe('unpublished')
  })

  it('refuses to unpublish a draft', () => {
    expect(() => anArticle().unpublish(editor)).toThrow(IllegalTransition)
  })
})

describe('isDueForPublication', () => {
  it('is due once the scheduled moment has passed', () => {
    const scheduled = anApprovedArticle({ status: 'scheduled', scheduledAt: EARLIER })
    expect(scheduled.isDueForPublication(NOW)).toBe(true)
  })

  it('is due exactly at the scheduled moment', () => {
    const scheduled = anApprovedArticle({ status: 'scheduled', scheduledAt: NOW })
    expect(scheduled.isDueForPublication(NOW)).toBe(true)
  })

  it('is not due before its moment', () => {
    const scheduled = anApprovedArticle({ status: 'scheduled', scheduledAt: LATER })
    expect(scheduled.isDueForPublication(NOW)).toBe(false)
  })

  it('is never due if it is not scheduled', () => {
    expect(anApprovedArticle().isDueForPublication(LATER)).toBe(false)
  })
})

describe('identity', () => {
  it('exposes the locale-scoped identity a reader URL is built from', () => {
    const article = anArticle()
    expect(article.id).toBe('art_1')
    expect(article.familyId).toBe('fam_1')
    expect(article.locale).toBe('en')
    expect(article.slug.value).toBe('budget-2026')
    expect(article.authorId).toBe(AUTHOR)
  })

  it('keeps the family shared across locales while slug and state diverge', () => {
    const en = anArticle()
    const fr = anArticle({ locale: 'fr', slug: Slug.of('budget-2026-fr'), status: 'in_review' })
    expect(fr.familyId).toBe(en.familyId)
    expect(fr.slug.equals(en.slug)).toBe(false)
    expect(fr.status).not.toBe(en.status)
  })
})

describe('immutability', () => {
  it('leaves the original article untouched', () => {
    const draft = anArticle()
    draft.submitForReview(author)
    expect(draft.status).toBe('draft')
  })
})

File: services/api/internal/domain/editorial/article_test.go
package editorial_test

import (
	"errors"
	"testing"
	"time"

	"github.com/kurasikapa/api/internal/domain/editorial"
	"github.com/kurasikapa/api/internal/domain/identity"
	"github.com/kurasikapa/api/internal/domain/shared"
)

var (
	authorID  = shared.UserID("usr_author")
	editorID  = shared.UserID("usr_editor")
	articleID = shared.ArticleID("art_1")
	revID     = shared.RevisionID("rev_1")
	now       = time.Date(2026, 8, 9, 12, 0, 0, 0, time.UTC)
)

func author() identity.Actor {
	return identity.NewActor(authorID, []identity.Role{identity.RoleAuthor})
}

func editor() identity.Actor {
	return identity.NewActor(editorID, []identity.Role{identity.RoleEditor})
}

func subscriber() identity.Actor {
	return identity.NewActor(shared.UserID("usr_reader"), []identity.Role{identity.RoleSubscriber})
}

func anArticle(mutate ...func(*editorial.ArticleState)) editorial.Article {
	slug, _ := shared.NewSlug("budget-2026")
	state := editorial.ArticleState{
		ID:       articleID,
		FamilyID: shared.FamilyID("fam_1"),
		Locale:   "en",
		Slug:     slug,
		Title:    "Budget 2026",
		AuthorID: authorID,
		Status:   editorial.StatusDraft,
	}

	for _, m := range mutate {
		m(&state)
	}

	return editorial.Reconstitute(state)
}

func TestSubmit(t *testing.T) {
	t.Parallel()

	t.Run("an author may submit their own draft", func(t *testing.T) {
		t.Parallel()

		got, err := anArticle().Submit(author())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if got.Status() != editorial.StatusInReview {
			t.Errorf("status = %s, want in_review", got.Status())
		}
	})

	t.Run("an author may not submit a colleague's draft", func(t *testing.T) {
		t.Parallel()

		other := identity.NewActor(shared.UserID("usr_other"), []identity.Role{identity.RoleAuthor})

		if _, err := anArticle().Submit(other); !errors.Is(err, editorial.ErrNotOwnArticle) {
			t.Errorf("got %v, want ErrNotOwnArticle", err)
		}
	})

	t.Run("submitting twice is refused", func(t *testing.T) {
		t.Parallel()

		submitted, err := anArticle().Submit(author())
		if err != nil {
			t.Fatalf("setup: %v", err)
		}
		if _, err := submitted.Submit(author()); !errors.Is(err, editorial.ErrIllegalTransition) {
			t.Errorf("got %v, want ErrIllegalTransition", err)
		}
	})
}

// The guard order is a confidentiality rule, not a style preference.
func TestGuardOrderDoesNotLeakAuthorship(t *testing.T) {
	t.Parallel()

	// A subscriber has no editorial permission at all. If ownership were
	// checked first they would be told the article "belongs to another
	// author" — which confirms both that an unpublished draft exists and that
	// someone else wrote it. Permission must be refused first.
	_, err := anArticle().Submit(subscriber())

	if !errors.Is(err, identity.ErrNotPermitted) {
		t.Fatalf("got %v, want ErrNotPermitted", err)
	}
	if errors.Is(err, editorial.ErrNotOwnArticle) {
		t.Error("leaked authorship to an actor with no editorial permission")
	}
}

func TestApprove(t *testing.T) {
	t.Parallel()

	inReview := func() editorial.Article {
		return anArticle(func(s *editorial.ArticleState) { s.Status = editorial.StatusInReview })
	}

	t.Run("an editor approves a named revision", func(t *testing.T) {
		t.Parallel()

		got, err := inReview().Approve(revID, articleID, editor())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		approved, ok := got.ApprovedRevisionID()
		if !ok || approved != revID {
			t.Errorf("approved revision = %v (ok=%v), want %s", approved, ok, revID)
		}
	})

	t.Run("an author may not approve their own work", func(t *testing.T) {
		t.Parallel()

		// The whole point of review. RoleAuthor has no article:approve.
		if _, err := inReview().Approve(revID, articleID, author()); !errors.Is(err, identity.ErrNotPermitted) {
			t.Errorf("got %v, want ErrNotPermitted", err)
		}
	})

	t.Run("refuses a revision belonging to a different article", func(t *testing.T) {
		t.Parallel()

		_, err := inReview().Approve(revID, shared.ArticleID("art_other"), editor())
		if !errors.Is(err, editorial.ErrRevisionNotOfArticle) {
			t.Errorf("got %v, want ErrRevisionNotOfArticle", err)
		}
	})
}

func TestRejectClearsTheApproval(t *testing.T) {
	t.Parallel()

	// Leaving the approval in place would let the next publish ship exactly
	// the text that was just refused.
	rejected, err := anArticle(func(s *editorial.ArticleState) {
		s.Status = editorial.StatusInReview
		s.ApprovedRevisionID = &revID
	}).Reject(editor())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if _, ok := rejected.ApprovedRevisionID(); ok {
		t.Error("approval survived a rejection")
	}
	if rejected.Status() != editorial.StatusDraft {
		t.Errorf("status = %s, want draft", rejected.Status())
	}
}

func TestPublish(t *testing.T) {
	t.Parallel()

	approved := func() editorial.Article {
		return anArticle(func(s *editorial.ArticleState) {
			s.Status = editorial.StatusApproved
			s.ApprovedRevisionID = &revID
		})
	}

	t.Run("publishes an approved article", func(t *testing.T) {
		t.Parallel()

		got, err := approved().Publish(now, editor())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if got.Status() != editorial.StatusPublished {
			t.Errorf("status = %s, want published", got.Status())
		}
		if at, ok := got.PublishedAt(); !ok || !at.Equal(now) {
			t.Errorf("publishedAt = %v (ok=%v), want %v", at, ok, now)
		}
	})

	t.Run("refuses to publish without an approved revision", func(t *testing.T) {
		t.Parallel()

		unapproved := anArticle(func(s *editorial.ArticleState) { s.Status = editorial.StatusApproved })

		if _, err := unapproved.Publish(now, editor()); !errors.Is(err, editorial.ErrNoApprovedRevision) {
			t.Errorf("got %v, want ErrNoApprovedRevision", err)
		}
	})

	t.Run("republishing does not restate the original publication date", func(t *testing.T) {
		t.Parallel()

		// publishedAt is the article's date of record. A story pulled for a
		// correction and put back must not resurface as today's news.
		first := now.Add(-72 * time.Hour)
		pulled := anArticle(func(s *editorial.ArticleState) {
			s.Status = editorial.StatusUnpublished
			s.ApprovedRevisionID = &revID
			s.PublishedAt = &first
		})

		got, err := pulled.Publish(now, editor())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if at, _ := got.PublishedAt(); !at.Equal(first) {
			t.Errorf("publishedAt = %v, want the original %v", at, first)
		}
	})
}

func TestSchedule(t *testing.T) {
	t.Parallel()

	approved := anArticle(func(s *editorial.ArticleState) {
		s.Status = editorial.StatusApproved
		s.ApprovedRevisionID = &revID
	})

	t.Run("schedules for a future time", func(t *testing.T) {
		t.Parallel()

		later := now.Add(time.Hour)

		got, err := approved.Schedule(later, now, editor())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if at, ok := got.ScheduledAt(); !ok || !at.Equal(later) {
			t.Errorf("scheduledAt = %v (ok=%v), want %v", at, ok, later)
		}
	})

	t.Run("refuses a time that has already passed", func(t *testing.T) {
		t.Parallel()

		_, err := approved.Schedule(now.Add(-time.Minute), now, editor())
		if !errors.Is(err, editorial.ErrScheduleInPast) {
			t.Errorf("got %v, want ErrScheduleInPast", err)
		}
	})
}

func TestRetitleFreezesTheSlugAfterPublication(t *testing.T) {
	t.Parallel()

	t.Run("an unpublished draft re-slugs with its new title", func(t *testing.T) {
		t.Parallel()

		got, err := anArticle().Retitle("Budget 2027 Explained", author())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if got.Slug().String() != "budget-2027-explained" {
			t.Errorf("slug = %q, want budget-2027-explained", got.Slug())
		}
	})

	t.Run("a once-published article keeps its address", func(t *testing.T) {
		t.Parallel()

		// Every inbound link, share and search result points at that URL.
		// The headline may still be corrected; the address may not move.
		published := now.Add(-time.Hour)
		pulled := anArticle(func(s *editorial.ArticleState) {
			s.Status = editorial.StatusUnpublished
			s.PublishedAt = &published
		})

		got, err := pulled.Retitle("Budget 2026: Corrected", author())
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if got.Slug().String() != "budget-2026" {
			t.Errorf("slug moved to %q after publication", got.Slug())
		}
		if got.Title() != "Budget 2026: Corrected" {
			t.Errorf("title = %q, want the correction to apply", got.Title())
		}
	})

	t.Run("refuses to edit an article that is in review", func(t *testing.T) {
		t.Parallel()

		inReview := anArticle(func(s *editorial.ArticleState) { s.Status = editorial.StatusInReview })

		if _, err := inReview.Retitle("Anything", author()); !errors.Is(err, editorial.ErrNotEditable) {
			t.Errorf("got %v, want ErrNotEditable", err)
		}
	})
}

func TestReadableBy(t *testing.T) {
	t.Parallel()

	draft := anArticle()
	published := anArticle(func(s *editorial.ArticleState) { s.Status = editorial.StatusPublished })

	cases := []struct {
		name    string
		article editorial.Article
		actor   identity.Actor
		want    bool
	}{
		{"anyone may read a published article", published, subscriber(), true},
		{"the author may read their own draft", draft, author(), true},
		{"an editor may read any draft", draft, editor(), true},
		{"a reader may not read someone's draft", draft, subscriber(), false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			if got := tc.article.ReadableBy(tc.actor); got != tc.want {
				t.Errorf("ReadableBy = %v, want %v", got, tc.want)
			}
		})
	}
}
